bfvmm/src/hve/arch/intel_x64/vmexit/cpuid.cpp

```diff
@@ -21,47 +21,144 @@
 
 #include <hve/arch/intel_x64/vcpu.h>
 
+void
+WEAK_SYM vcpu_init_root(vcpu_t *vcpu)
+{ bfignored(vcpu); }
+
+void
+WEAK_SYM vcpu_fini_root(vcpu_t *vcpu)
+{ bfignored(vcpu); }
+
 namespace bfvmm::intel_x64
 {
 
 static bool
-handle_cpuid_feature_information(
-    vcpu_t vcpu, cpuid_handler::info_t &info)
+handle_cpuid_feature_information(vcpu *vcpu)
 {
-    bfignored(vcpu);
+    using namespace ::intel_x64::cpuid;
 
     // Currently, we do not support nested virtualization. As a result,
     // the EAPIs adds a default handler to disable support for VMXE here.
     //
 
-    info.rcx =
-        clear_bit(
-            info.rcx, ::intel_x64::cpuid::feature_information::ecx::vmx::from
-        );
+    vcpu->set_rcx(
+        clear_bit(vcpu->rcx(), feature_information::ecx::vmx::from)
+    );
+
+    return false;
+}
+
+static bool
+handle_cpuid_0x4BF00000(vcpu *vcpu)
+{
+    /// Ack
+    ///
+    /// This can be used by an application to ack the existence of the
+    /// hypervisor. This is useful because vmcall only exists if the hypervisor
+    /// is running while cpuid can be run from any ring, and always exists
+    /// which means it can be used to ack safely from any application.
+    ///
+
+    vcpu->set_rax(0x4BF00001);
+    return vcpu->advance();
+}
 
-    return true;
+static bool
+handle_cpuid_0x4BF00010(vcpu *vcpu)
+{
+    /// Init
+    ///
+    /// Some initialization is required after the hypervisor has started. For
+    /// example, any memory mapped resources such as ACPI or VT-d need to be
+    /// initalized using the VMM's CR3, and not the hosts.
+    ///
+
+    vcpu_init_root(vcpu);
+    return vcpu->advance();
+}
+
+static bool
+handle_cpuid_0x4BF00011(vcpu *vcpu)
```

Comment:
Can we start a list of constexprs or defines of the bareflank-specific cpuids rather than using the hex value?

Reply:
yes. Let's do that in another PR, but I agree... for the BF specific leaves, this should be English, not in hex. For hardware that actually exists I prefer the hex as you have to know the hex anyways (the English version for Intel doesn't mean much).

Reply:
>  this should be English, not in hex.
:+1: 

> For hardware that actually exists I prefer the hex as you have to know the hex anyways
Can you give an example of this? You said hardware; what is the actual scope you're referring to? What specific registers/devices/etc?